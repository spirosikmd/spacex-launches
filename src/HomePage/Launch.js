import React from 'react';
import PropTypes from 'prop-types';
import { Link } from '@reach/router';
import { withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import CardActions from '@material-ui/core/CardActions';
import LaunchDateTime from '../LaunchDateTime';
import LaunchStatus from '../LaunchStatus';

const styles = theme => ({
  launch: {
    display: 'flex',
    alignItems: 'center',
  },
  launchInfo: {
    flexGrow: 1,
  },
  connector: {
    minWidth: theme.spacing.unit * 2,
    border: `0.5px dashed ${theme.palette.secondary.dark}`,
    marginTop: '1px',
  },
});

const Launch = ({ classes, launch }) => {
  const hadMissionIds = launch.missionIds.length > 0;
  const hasContent = launch.details || hadMissionIds;

  return (
    <div className={classes.launch}>
      <LaunchStatus
        isSuccessful={launch.isSuccessful}
        isFailed={launch.isFailed}
        isUpcoming={launch.isUpcoming}
        isInProgress={launch.isInProgress}
      />
      <div className={classes.connector} />
      <Card className={classes.launchInfo}>
        <CardHeader
          title={launch.missionName}
          subheader={
            <LaunchDateTime
              utcDate={launch.utcDate}
              isTentative={launch.isTentative}
            />
          }
        />
        {hasContent && (
          <CardContent>
            {launch.details &&
              (hadMissionIds ? (
                <Typography component="p" gutterBottom>
                  {launch.details}
                </Typography>
              ) : (
                <Typography component="p">{launch.details}</Typography>
              ))}
            {hadMissionIds && (
              <Typography color="textSecondary" component="p">
                ID: {launch.missionIds.join(', ')}
              </Typography>
            )}
          </CardContent>
        )}
        {!launch.isTentative && (
          <CardActions>
            <Button
              size="small"
              color="primary"
              component={Link}
              to={`/launches/${launch.flightNumber}`}
            >
              explore
            </Button>
          </CardActions>
        )}
      </Card>
    </div>
  );
};

export const LaunchPropType = {
  isSuccessful: PropTypes.bool.isRequired,
  isFailed: PropTypes.bool.isRequired,
  isUpcoming: PropTypes.bool.isRequired,
  isInProgress: PropTypes.bool.isRequired,
  isTentative: PropTypes.bool.isRequired,
  missionName: PropTypes.string.isRequired,
  utcDate: PropTypes.object.isRequired,
  details: PropTypes.string,
  missionIds: PropTypes.array.isRequired,
  flightNumber: PropTypes.number.isRequired,
};

Launch.propTypes = {
  classes: PropTypes.object.isRequired,
  launch: PropTypes.shape(LaunchPropType).isRequired,
};

export default withStyles(styles)(React.memo(Launch));
