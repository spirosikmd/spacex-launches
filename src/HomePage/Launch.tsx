import React from 'react';
import PropTypes from 'prop-types';
import { Link } from '@reach/router';
import {
  withStyles,
  Theme,
  WithStyles,
  createStyles,
} from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import Button, { ButtonProps } from '@material-ui/core/Button';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import CardActions from '@material-ui/core/CardActions';
import LaunchDateTime from '../LaunchDateTime';
import LaunchStatus from '../LaunchStatus';
import LaunchMissionIds from '../LaunchMissionIds';
import { LaunchData } from '../api';

const styles = (theme: Theme) =>
  createStyles({
    launch: {
      display: 'flex',
      alignItems: 'center',
    },
    launchInfo: {
      flexGrow: 1,
    },
    connector: {
      minWidth: theme.spacing(2),
      border: `0.5px dashed ${theme.palette.secondary.dark}`,
      marginTop: '1px',
    },
  });

interface LinkButtonProps extends ButtonProps {
  to: string;
}

const LinkButton = (props: LinkButtonProps) => (
  <Button {...props} component={Link as any} />
);

interface LaunchProps extends WithStyles<typeof styles> {
  launch: LaunchData;
}

const Launch = ({ classes, launch }: LaunchProps) => {
  const hadMissionIds = launch.missionIds.length > 0;
  const hasContent = launch.details || hadMissionIds;

  return (
    <div className={classes.launch}>
      <LaunchStatus
        isSuccessful={launch.isSuccessful}
        isFailed={launch.isFailed}
        isUpcoming={launch.isUpcoming}
        isInProgress={launch.isInProgress}
      />
      <div className={classes.connector} />
      <Card className={classes.launchInfo}>
        <CardHeader
          title={launch.missionName}
          subheader={
            <LaunchDateTime
              utcDate={launch.utcDate}
              isTentative={launch.isTentative}
            />
          }
        />
        {hasContent && (
          <CardContent>
            {launch.details &&
              (hadMissionIds ? (
                <Typography component="p" gutterBottom>
                  {launch.details}
                </Typography>
              ) : (
                <Typography component="p">{launch.details}</Typography>
              ))}
            {hadMissionIds && (
              <LaunchMissionIds missionIds={launch.missionIds} />
            )}
          </CardContent>
        )}
        {!launch.isTentative && (
          <CardActions>
            <LinkButton
              size="small"
              color="primary"
              to={`/launches/${launch.flightNumber}`}
            >
              explore
            </LinkButton>
          </CardActions>
        )}
      </Card>
    </div>
  );
};

export const LaunchPropType = {
  isSuccessful: PropTypes.bool.isRequired,
  isFailed: PropTypes.bool.isRequired,
  isUpcoming: PropTypes.bool.isRequired,
  isInProgress: PropTypes.bool.isRequired,
  isTentative: PropTypes.bool.isRequired,
  missionName: PropTypes.string.isRequired,
  utcDate: PropTypes.object.isRequired,
  details: PropTypes.string,
  missionIds: PropTypes.array.isRequired,
  flightNumber: PropTypes.number.isRequired,
};

Launch.propTypes = {
  classes: PropTypes.object.isRequired,
  launch: PropTypes.shape(LaunchPropType).isRequired,
};

export default withStyles(styles)(React.memo(Launch));
