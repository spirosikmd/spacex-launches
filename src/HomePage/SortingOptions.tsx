import React, { SyntheticEvent } from 'react';
import PropTypes from 'prop-types';
import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import { ASC, DESC, FLIGHT_NUMBER_FIELD, UTC_DATE_FIELD } from '../constants';

function getSortOrderOptions(sortField: string) {
  switch (sortField) {
    case FLIGHT_NUMBER_FIELD:
      return (
        <>
          <option value={ASC}>smaller to bigger</option>
          <option value={DESC}>bigger to smaller</option>
        </>
      );
    case UTC_DATE_FIELD:
      return (
        <>
          <option value={ASC}>old to new</option>
          <option value={DESC}>new to old</option>
        </>
      );
    default:
      return (
        <>
          <option value={ASC}>ascending</option>
          <option value={DESC}>descending</option>
        </>
      );
  }
}

interface SortingOptionsProps {
  sortField: string;
  sortOrder: string;
  onSortChange: (name: string, value: string) => void;
}

const SortingOptions = ({
  sortField,
  sortOrder,
  onSortChange,
}: SortingOptionsProps) => {
  const handleSortChange = (event: SyntheticEvent) => {
    const target = event.currentTarget as HTMLSelectElement;
    onSortChange(target.name, target.value);
  };

  return (
    <Grid container spacing={8}>
      <Grid item>
        <TextField
          id="sort-field"
          select
          label="Sort field"
          value={sortField}
          onChange={handleSortChange}
          SelectProps={{
            native: true,
            name: 'sortField',
          }}
          margin="normal"
          variant="outlined"
        >
          <option value={FLIGHT_NUMBER_FIELD}>flight number</option>
          <option value={UTC_DATE_FIELD}>utc date</option>
        </TextField>
      </Grid>
      <Grid item>
        <TextField
          id="sort-order"
          select
          label="Sort order"
          value={sortOrder}
          onChange={handleSortChange}
          SelectProps={{
            native: true,
            name: 'sortOrder',
          }}
          margin="normal"
          variant="outlined"
        >
          {getSortOrderOptions(sortField)}
        </TextField>
      </Grid>
    </Grid>
  );
};

SortingOptions.propTypes = {
  sortField: PropTypes.string.isRequired,
  sortOrder: PropTypes.string.isRequired,
  onSortChange: PropTypes.func.isRequired,
};

export default React.memo(SortingOptions);
