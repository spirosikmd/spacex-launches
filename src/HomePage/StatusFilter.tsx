import React, { SyntheticEvent } from 'react';
import PropTypes from 'prop-types';
import FormControl from '@material-ui/core/FormControl';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import FormGroup from '@material-ui/core/FormGroup';
import FormLabel from '@material-ui/core/FormLabel';
import Switch from '@material-ui/core/Switch';

interface StatusFilterProps {
  showFailed: boolean;
  showSuccessful: boolean;
  showUpcoming: boolean;
  onChange: (value: string, checked: boolean) => void;
}

const StatusFilter = ({
  showFailed,
  showSuccessful,
  showUpcoming,
  onChange,
}: StatusFilterProps) => {
  const handleChange = (event: SyntheticEvent) => {
    const target = event.currentTarget as HTMLInputElement;
    onChange(target.value, target.checked);
  };

  return (
    <FormControl component={'fieldset' as 'div'}>
      <FormLabel component={'legend' as 'label'}>Show launches</FormLabel>
      <FormGroup row>
        <FormControlLabel
          control={
            <Switch
              checked={showFailed}
              onChange={handleChange}
              value="showFailed"
            />
          }
          label="Failed"
        />
        <FormControlLabel
          control={
            <Switch
              checked={showSuccessful}
              onChange={handleChange}
              value="showSuccessful"
            />
          }
          label="Successful"
        />
        <FormControlLabel
          control={
            <Switch
              checked={showUpcoming}
              onChange={handleChange}
              value="showUpcoming"
            />
          }
          label="Upcoming"
        />
      </FormGroup>
    </FormControl>
  );
};

StatusFilter.propTypes = {
  showFailed: PropTypes.bool.isRequired,
  showSuccessful: PropTypes.bool.isRequired,
  showUpcoming: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
};

export default React.memo(StatusFilter);
