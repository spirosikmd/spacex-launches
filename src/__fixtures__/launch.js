export function createLaunchResponse(data) {
  return {
    flight_number: 1,
    mission_name: 'Test',
    launch_date_utc: '2019-01-01T18:16:00.000Z',
    launch_success: true,
    upcoming: false,
    is_tentative: false,
    details: 'Details',
    mission_id: ['id1'],
    links: {
      mission_patch: 'missionPatch',
    },
    ...data,
  };
}

export function createLaunch(data) {
  return {
    flightNumber: 1,
    missionName: 'Test',
    utcDate: new Date('2019-01-01'),
    isSuccessful: true,
    isFailed: false,
    isUpcoming: false,
    isInProgress: false,
    isTentative: false,
    details: 'Details',
    missionIds: ['id1'],
    missionPatch: 'missionPatch',
    ...data,
  };
}
