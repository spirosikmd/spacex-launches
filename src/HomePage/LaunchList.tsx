import React from 'react';
import PropTypes from 'prop-types';
import {
  withStyles,
  Theme,
  WithStyles,
  createStyles,
} from '@material-ui/core/styles';
import Launch, { LaunchPropType } from './Launch';
import Image from '../Image';
import emptyListImage from './emptyList.png';
import { LaunchData } from '../api';

const styles = (theme: Theme) =>
  createStyles({
    launchList: {
      display: 'flex',
      flexDirection: 'column' as 'column',
      position: 'relative' as 'relative',
    },
    timeline: {
      borderLeft: `1px solid ${theme.palette.grey[300]}`,
      position: 'absolute' as 'absolute',
      top: '6px',
      bottom: '6px',
      left: `${theme.spacing.unit + 4}px`,
      '&::before': {
        position: 'absolute' as 'absolute',
        content: '""',
        borderLeft: `1px dashed ${theme.palette.grey[300]}`,
        top: '-13px',
        left: '-1px',
        height: '12px',
      },
      '&::after': {
        position: 'absolute' as 'absolute',
        content: '""',
        borderLeft: `1px dashed ${theme.palette.grey[300]}`,
        bottom: '-13px',
        left: '-1px',
        height: '12px',
      },
    },
    launchItem: {
      padding: `${theme.spacing.unit * 2}px 0`,
    },
  });

interface LaunchListProps extends WithStyles<typeof styles> {
  launches: LaunchData[];
}

const LaunchList = ({ classes, launches }: LaunchListProps) => {
  if (launches.length === 0) {
    return <Image src={emptyListImage} alt="No launches..." />;
  }

  return (
    <div className={classes.launchList}>
      <div className={classes.timeline} />
      {launches.map(launch => (
        <div key={launch.flightNumber} className={classes.launchItem}>
          <Launch launch={launch} />
        </div>
      ))}
    </div>
  );
};

LaunchList.propTypes = {
  classes: PropTypes.object.isRequired,
  launches: PropTypes.arrayOf(PropTypes.shape(LaunchPropType)).isRequired,
};

export default withStyles(styles)(React.memo(LaunchList));
