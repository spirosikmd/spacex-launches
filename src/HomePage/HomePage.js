import React, { PureComponent } from 'react';
import { withStyles } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import GeneralInfo from '../GeneralInfo';
import LaunchList from '../LaunchList';
import SortingOptions from '../SortingOptions';
import StatusFilter from '../StatusFilter';
import Loader from '../Loader';
import { getLaunches } from '../api';
import { processLaunches } from '../utils';
import { DESC, UTC_DATE_FIELD, FIELDS } from '../constants';

const styles = theme => ({
  loader: {
    paddingTop: theme.spacing.unit * 8,
    paddingBottom: theme.spacing.unit * 8,
  },
});

export class HomePage extends PureComponent {
  state = {
    launches: [],
    isLoadingLaunches: true,
    isUpdatingLaunches: false,
    error: null,
    showUpcoming: false,
    showSuccessful: true,
    showFailed: true,
    sortOrder: DESC,
    sortField: UTC_DATE_FIELD,
  };

  async componentDidMount() {
    try {
      const { sortField, sortOrder } = this.state;
      const launches = await getLaunches({
        sortField,
        sortOrder,
        filter: FIELDS,
      });
      this.setState({ launches, isLoadingLaunches: false });
    } catch (error) {
      this.setState({ error, isLoadingLaunches: false });
    }
  }

  handleChange = (value, checked) => {
    this.setState({ [value]: checked });
  };

  handleSortChange = (name, value) => {
    this.setState(
      {
        [name]: value,
        isUpdatingLaunches: true,
      },
      async () => {
        const { sortField, sortOrder } = this.state;
        const launches = await getLaunches({
          sortField,
          sortOrder,
          filter: FIELDS,
        });
        this.setState({ launches, isUpdatingLaunches: false });
      }
    );
  };

  render() {
    const { classes } = this.props;
    const {
      launches,
      isLoadingLaunches,
      isUpdatingLaunches,
      showUpcoming,
      showSuccessful,
      showFailed,
      sortOrder,
      sortField,
    } = this.state;

    const processedLaunches = processLaunches(launches, {
      showUpcoming,
      showSuccessful,
      showFailed,
    });

    if (isLoadingLaunches) {
      return (
        <div className={classes.loader}>
          <Loader />
        </div>
      );
    }

    return (
      <Grid container spacing={16}>
        <Grid item xs={12}>
          <GeneralInfo launches={launches} />
        </Grid>
        <Grid item xs={12}>
          <StatusFilter
            showFailed={showFailed}
            showSuccessful={showSuccessful}
            showUpcoming={showUpcoming}
            onChange={this.handleChange}
          />
        </Grid>
        {processedLaunches.length > 0 && (
          <Grid item xs={12}>
            <SortingOptions
              sortField={sortField}
              sortOrder={sortOrder}
              onSortChange={this.handleSortChange}
            />
          </Grid>
        )}
        <Grid item xs={12}>
          {isUpdatingLaunches ? (
            <div className={classes.loader}>
              <Loader />
            </div>
          ) : (
            <LaunchList launches={processedLaunches} />
          )}
        </Grid>
      </Grid>
    );
  }
}

export default withStyles(styles)(React.memo(HomePage));
