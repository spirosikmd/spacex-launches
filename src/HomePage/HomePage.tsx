import React, { PureComponent } from 'react';
import {
  withStyles,
  Theme,
  createStyles,
  WithStyles,
} from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import GeneralInfo from './GeneralInfo';
import LaunchList from './LaunchList';
import SortingOptions from './SortingOptions';
import StatusFilter from './StatusFilter';
import Loader from '../Loader';
import { getLaunches, LaunchData } from '../api';
import { DESC, UTC_DATE_FIELD } from '../constants';
import { RouteComponentProps, WindowLocation } from '@reach/router';

function processLaunches(
  launches: LaunchData[],
  {
    showUpcoming,
    showSuccessful,
    showFailed,
  }: { showUpcoming: boolean; showSuccessful: boolean; showFailed: boolean }
) {
  let processedLaunches = [...launches];

  if (!showUpcoming) {
    processedLaunches = processedLaunches.filter(launch => !launch.isUpcoming);
  }

  if (!showSuccessful) {
    processedLaunches = processedLaunches.filter(
      launch => !launch.isSuccessful
    );
  }

  if (!showFailed) {
    processedLaunches = processedLaunches.filter(launch => !launch.isFailed);
  }

  return processedLaunches;
}

function getBooleanParam(param: string | null) {
  return param === null ? true : param === 'true';
}

function getParams(
  location: WindowLocation | undefined
): {
  showUpcoming: boolean;
  showSuccessful: boolean;
  showFailed: boolean;
  sortOrder: string;
  sortField: string;
} {
  if (!location) {
    return {
      showUpcoming: true,
      showSuccessful: true,
      showFailed: true,
      sortOrder: DESC,
      sortField: UTC_DATE_FIELD,
    };
  }

  const searchParams = new URLSearchParams(location.search);

  return {
    showUpcoming: getBooleanParam(searchParams.get('upcoming')),
    showSuccessful: getBooleanParam(searchParams.get('successful')),
    showFailed: getBooleanParam(searchParams.get('failed')),
    sortOrder: searchParams.get('order') || DESC,
    sortField: searchParams.get('sort') || UTC_DATE_FIELD,
  };
}

function setParams({
  showUpcoming = true,
  showSuccessful = true,
  showFailed = true,
  sortOrder = DESC,
  sortField = UTC_DATE_FIELD,
}) {
  const searchParams = new URLSearchParams();
  searchParams.set('upcoming', showUpcoming.toString());
  searchParams.set('successful', showSuccessful.toString());
  searchParams.set('failed', showFailed.toString());
  searchParams.set('order', sortOrder);
  searchParams.set('sort', sortField);
  return searchParams.toString();
}

const styles = (theme: Theme) =>
  createStyles({
    loader: {
      paddingTop: theme.spacing.unit * 8,
      paddingBottom: theme.spacing.unit * 8,
    },
  });

interface HomePageState {
  launches: LaunchData[];
  isLoadingLaunches: boolean;
  isUpdatingLaunches: boolean;
  error: Error | null;
}

interface HomePageProps
  extends RouteComponentProps,
    WithStyles<typeof styles> {}

class HomePage extends PureComponent<HomePageProps, HomePageState> {
  state: HomePageState = {
    launches: [],
    isLoadingLaunches: true,
    isUpdatingLaunches: false,
    error: null,
  };

  async componentDidMount() {
    try {
      const { location } = this.props;
      const { sortField, sortOrder } = getParams(location);
      const launches = await getLaunches({
        sortField,
        sortOrder,
      });
      this.setState({ launches, isLoadingLaunches: false });
    } catch (error) {
      this.setState({ error, isLoadingLaunches: false });
    }

    this.navigateWithParams();
  }

  navigateWithParams(newParams = {}) {
    const { location, navigate } = this.props;
    if (!navigate) return;
    const params = getParams(location);
    const url = setParams({ ...params, ...newParams });
    navigate(`?${url}`);
  }

  handleStatusFilterChange = (value: string, checked: boolean) => {
    this.navigateWithParams({ [value]: checked });
  };

  handleSortChange = async (name: string, value: string) => {
    const { location } = this.props;
    this.setState({ isUpdatingLaunches: true });
    this.navigateWithParams({ [name]: value });
    const { sortField, sortOrder } = getParams(location);
    const launches = await getLaunches({
      sortField,
      sortOrder,
      ...{ [name]: value },
    });
    this.setState({ launches, isUpdatingLaunches: false });
  };

  render() {
    const { classes, location } = this.props;
    const {
      launches,
      isLoadingLaunches,
      isUpdatingLaunches,
      error,
    } = this.state;

    const {
      showUpcoming,
      showSuccessful,
      showFailed,
      sortField,
      sortOrder,
    } = getParams(location);

    const processedLaunches = processLaunches(launches, {
      showUpcoming,
      showSuccessful,
      showFailed,
    });

    if (isLoadingLaunches) {
      return (
        <div className={classes.loader}>
          <Loader />
        </div>
      );
    }

    if (error) {
      return <Typography>{error.message}</Typography>;
    }

    return (
      <Grid container spacing={16}>
        <Grid item xs={12}>
          <GeneralInfo launches={launches} />
        </Grid>
        <Grid item xs={12}>
          <StatusFilter
            showFailed={showFailed}
            showSuccessful={showSuccessful}
            showUpcoming={showUpcoming}
            onChange={this.handleStatusFilterChange}
          />
        </Grid>
        {processedLaunches.length > 0 && (
          <Grid item xs={12}>
            <SortingOptions
              sortField={sortField}
              sortOrder={sortOrder}
              onSortChange={this.handleSortChange}
            />
          </Grid>
        )}
        <Grid item xs={12}>
          {isUpdatingLaunches ? (
            <div className={classes.loader}>
              <Loader />
            </div>
          ) : (
            <LaunchList launches={processedLaunches} />
          )}
        </Grid>
      </Grid>
    );
  }
}

export default withStyles(styles)(React.memo(HomePage));
