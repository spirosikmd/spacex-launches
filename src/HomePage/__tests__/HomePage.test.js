import HomePage from '../HomePage';
import { createLaunch } from '../../__fixtures__/launch';
import { getLaunches } from '../../api';
import { FLIGHT_NUMBER_FIELD, DESC } from '../../constants';
import StatusFilter from '../StatusFilter';
import SortingOptions from '../SortingOptions';

jest.mock('../../api', () => ({
  getLaunches: jest.fn(),
}));

describe('HomePage', () => {
  let props;

  beforeEach(() => {
    props = {
      classes: {
        loader: 'loader',
      },
      location: {
        search:
          'failed=true&successful=true&upcoming=false&sortOrder=desc&sortField=utcDate',
      },
      navigate: jest.fn(),
    };
    const launches = [
      createLaunch(),
      createLaunch({ isSuccessful: false, isFailed: true }),
      createLaunch({ isSuccessful: false, isUpcoming: true }),
    ];
    getLaunches.mockReturnValue(Promise.resolve(launches));
  });

  it('renders', async () => {
    const homePage = shallow(HomePage, props);
    await Promise.resolve();
    homePage.update();
    expect(homePage).toMatchSnapshot();
    expect(props.navigate).toBeCalledWith(
      '?upcoming=false&successful=true&failed=true&order=desc&sort=utcDate'
    );
  });

  describe('when is loading launches', () => {
    it('renders a loader', () => {
      expect(shallow(HomePage, props)).toMatchSnapshot();
    });
  });

  describe('when is updating launches', () => {
    it('renders info, filters, sorting options, and a loader', async () => {
      const homePage = shallow(HomePage, props);
      await Promise.resolve();
      homePage.update();
      const sortingOptions = homePage.find(SortingOptions);
      sortingOptions.prop('onSortChange')('sortField', FLIGHT_NUMBER_FIELD);
      expect(homePage).toMatchSnapshot();
      expect(props.navigate).toBeCalledWith(
        '?upcoming=false&successful=true&failed=true&order=desc&sort=flightNumber'
      );
    });
  });

  describe('when there are no processed launches', () => {
    it('does not render the sorting options', async () => {
      props.location.search =
        'failed=false&successful=false&upcoming=false&sortOrder=desc&sortField=utcDate';
      const homePage = shallow(HomePage, props);
      await Promise.resolve();
      homePage.update();
      expect(homePage).toMatchSnapshot();
    });
  });

  describe('when any status filter changes', () => {
    it('navigates with the new value and checked', async () => {
      const homePage = shallow(HomePage, props);
      await Promise.resolve();
      homePage.update();
      const statusFilter = homePage.find(StatusFilter);
      statusFilter.prop('onChange')('showUpcoming', 'true');
      expect(props.navigate).toBeCalledWith(
        '?upcoming=true&successful=true&failed=true&order=desc&sort=utcDate'
      );
    });
  });

  describe('when sorting options change', () => {
    it('gets launches with new sorting options', async () => {
      const homePage = shallow(HomePage, props);
      await Promise.resolve();
      homePage.update();
      const sortingOptions = homePage.find(SortingOptions);
      await sortingOptions.prop('onSortChange')(
        'sortField',
        FLIGHT_NUMBER_FIELD
      );
      const state = homePage.state();
      expect(state.isUpdatingLaunches).toBe(false);
      expect(getLaunches).toBeCalledWith({
        sortField: FLIGHT_NUMBER_FIELD,
        sortOrder: DESC,
      });
    });
  });
});
