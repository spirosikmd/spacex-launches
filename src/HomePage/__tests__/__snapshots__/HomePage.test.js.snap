// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage renders 1`] = `
<WithStyles(Grid)
  container={true}
  spacing={16}
>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <React.Memo
      launches={
        Array [
          Object {
            "details": "Details",
            "flightNumber": 1,
            "isFailed": false,
            "isInProgress": false,
            "isSuccessful": true,
            "isTentative": false,
            "isUpcoming": false,
            "missionIds": Array [
              "id1",
            ],
            "missionName": "Test",
            "utcDate": Object {
              "getDate": [MockFunction],
              "getFullYear": [MockFunction],
              "getMonth": [MockFunction],
              "toLocaleString": [MockFunction],
            },
          },
          Object {
            "details": "Details",
            "flightNumber": 1,
            "isFailed": true,
            "isInProgress": false,
            "isSuccessful": false,
            "isTentative": false,
            "isUpcoming": false,
            "missionIds": Array [
              "id1",
            ],
            "missionName": "Test",
            "utcDate": Object {
              "getDate": [MockFunction],
              "getFullYear": [MockFunction],
              "getMonth": [MockFunction],
              "toLocaleString": [MockFunction],
            },
          },
          Object {
            "details": "Details",
            "flightNumber": 1,
            "isFailed": false,
            "isInProgress": false,
            "isSuccessful": false,
            "isTentative": false,
            "isUpcoming": true,
            "missionIds": Array [
              "id1",
            ],
            "missionName": "Test",
            "utcDate": Object {
              "getDate": [MockFunction],
              "getFullYear": [MockFunction],
              "getMonth": [MockFunction],
              "toLocaleString": [MockFunction],
            },
          },
        ]
      }
    />
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <React.Memo
      onChange={[Function]}
      showFailed={true}
      showSuccessful={true}
      showUpcoming={false}
    />
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <React.Memo
      onSortChange={[Function]}
      sortField="launch_date_utc"
      sortOrder="desc"
    />
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <WithStyles(Component)
      launches={
        Array [
          Object {
            "details": "Details",
            "flightNumber": 1,
            "isFailed": false,
            "isInProgress": false,
            "isSuccessful": true,
            "isTentative": false,
            "isUpcoming": false,
            "missionIds": Array [
              "id1",
            ],
            "missionName": "Test",
            "utcDate": Object {
              "getDate": [MockFunction],
              "getFullYear": [MockFunction],
              "getMonth": [MockFunction],
              "toLocaleString": [MockFunction],
            },
          },
          Object {
            "details": "Details",
            "flightNumber": 1,
            "isFailed": true,
            "isInProgress": false,
            "isSuccessful": false,
            "isTentative": false,
            "isUpcoming": false,
            "missionIds": Array [
              "id1",
            ],
            "missionName": "Test",
            "utcDate": Object {
              "getDate": [MockFunction],
              "getFullYear": [MockFunction],
              "getMonth": [MockFunction],
              "toLocaleString": [MockFunction],
            },
          },
        ]
      }
    />
  </WithStyles(Grid)>
</WithStyles(Grid)>
`;

exports[`HomePage when is loading launches renders a loader 1`] = `
<div
  className="loader"
>
  <WithStyles(Component) />
</div>
`;

exports[`HomePage when is updating launches renders info, filters, sorting options, and a loader 1`] = `
<WithStyles(Grid)
  container={true}
  spacing={16}
>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <React.Memo
      launches={
        Array [
          Object {
            "details": "Details",
            "flightNumber": 1,
            "isFailed": false,
            "isInProgress": false,
            "isSuccessful": true,
            "isTentative": false,
            "isUpcoming": false,
            "missionIds": Array [
              "id1",
            ],
            "missionName": "Test",
            "utcDate": Object {
              "getDate": [MockFunction],
              "getFullYear": [MockFunction],
              "getMonth": [MockFunction],
              "toLocaleString": [MockFunction],
            },
          },
          Object {
            "details": "Details",
            "flightNumber": 1,
            "isFailed": true,
            "isInProgress": false,
            "isSuccessful": false,
            "isTentative": false,
            "isUpcoming": false,
            "missionIds": Array [
              "id1",
            ],
            "missionName": "Test",
            "utcDate": Object {
              "getDate": [MockFunction],
              "getFullYear": [MockFunction],
              "getMonth": [MockFunction],
              "toLocaleString": [MockFunction],
            },
          },
          Object {
            "details": "Details",
            "flightNumber": 1,
            "isFailed": false,
            "isInProgress": false,
            "isSuccessful": false,
            "isTentative": false,
            "isUpcoming": true,
            "missionIds": Array [
              "id1",
            ],
            "missionName": "Test",
            "utcDate": Object {
              "getDate": [MockFunction],
              "getFullYear": [MockFunction],
              "getMonth": [MockFunction],
              "toLocaleString": [MockFunction],
            },
          },
        ]
      }
    />
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <React.Memo
      onChange={[Function]}
      showFailed={true}
      showSuccessful={true}
      showUpcoming={false}
    />
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <React.Memo
      onSortChange={[Function]}
      sortField="flight_number"
      sortOrder="desc"
    />
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <div
      className="loader"
    >
      <WithStyles(Component) />
    </div>
  </WithStyles(Grid)>
</WithStyles(Grid)>
`;

exports[`HomePage when there are no processed launches does not render the sorting options 1`] = `
<WithStyles(Grid)
  container={true}
  spacing={16}
>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <React.Memo
      launches={
        Array [
          Object {
            "details": "Details",
            "flightNumber": 1,
            "isFailed": false,
            "isInProgress": false,
            "isSuccessful": true,
            "isTentative": false,
            "isUpcoming": false,
            "missionIds": Array [
              "id1",
            ],
            "missionName": "Test",
            "utcDate": Object {
              "getDate": [MockFunction],
              "getFullYear": [MockFunction],
              "getMonth": [MockFunction],
              "toLocaleString": [MockFunction],
            },
          },
          Object {
            "details": "Details",
            "flightNumber": 1,
            "isFailed": true,
            "isInProgress": false,
            "isSuccessful": false,
            "isTentative": false,
            "isUpcoming": false,
            "missionIds": Array [
              "id1",
            ],
            "missionName": "Test",
            "utcDate": Object {
              "getDate": [MockFunction],
              "getFullYear": [MockFunction],
              "getMonth": [MockFunction],
              "toLocaleString": [MockFunction],
            },
          },
          Object {
            "details": "Details",
            "flightNumber": 1,
            "isFailed": false,
            "isInProgress": false,
            "isSuccessful": false,
            "isTentative": false,
            "isUpcoming": true,
            "missionIds": Array [
              "id1",
            ],
            "missionName": "Test",
            "utcDate": Object {
              "getDate": [MockFunction],
              "getFullYear": [MockFunction],
              "getMonth": [MockFunction],
              "toLocaleString": [MockFunction],
            },
          },
        ]
      }
    />
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <React.Memo
      onChange={[Function]}
      showFailed={true}
      showSuccessful={true}
      showUpcoming={false}
    />
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={12}
  >
    <WithStyles(Component)
      launches={Array []}
    />
  </WithStyles(Grid)>
</WithStyles(Grid)>
`;
